torch/csrc/dynamo/guards.cpp

```diff
@@ -574,6 +574,459 @@ static struct PyModuleDef _module = {
+  py::function _print_failure_fn;
+};
+
+/**
+ * This is an example of relational guard and checks tensor X is not tensor Y.
+ * We construct one leaf guard and install it at as a leaf of two guard managers
+ * (one for X and another for Y). Therefore, this guard is run twice. In the
+ * first invocation, it saves the first value and returns True. In the second
+ * invocation, it compares the saved value with the new value and returns True
+ * if they do not alias.
+ */
+class NoTensorAliasingGuard : public LeafGuard {
+ public:
+  NoTensorAliasingGuard() {
+    _is_first_call = true;
+  }
+
+  bool check_nopybind(PyObject* value) override { // borrowed ref
+    if (_is_first_call) {
+      _first_tensor = value;
+      _is_first_call = false;
+      return true;
+    }
+    _is_first_call = true;
+    return _first_tensor != value;
+  }
+
+  std::string get_failure_reason(PyObject* value) override { // borrowed ref
+    return "";
+  }
+
+ private:
+  bool _is_first_call;
+  PyObject* _first_tensor;
+};
+
+class GuardManager;
+/**
+ * Base class representing a pair of accessor and the associated guard manager.
+ * The accessor defines how to access the child value from the py::object given
+ * to the parent check function.
+ *
+ * GuardAccessors can be considered equivalent to name() method of Source
+ * objects in guards.py. In python, name() method returns a str which we can
+ * then eval in f_locals and f_globals to retrieve the actual py object.
+ * GuardAccessor serves the same purpose. The minor difference is that
+ * GuardManager is a tree structure, so a GuardAccessor just has to retrieve the
+ * value in the next level in this tree and pass it to the child GuardAccessor.
+ *
+ * GuardAccessor also owns the GuardManager associated with the retrieved value
+ * from the GuardAccessor.
+ */
+class GuardAccessor {
+ public:
+  GuardAccessor(py::object accessor_key)
+      : _guard_manager(std::make_unique<GuardManager>()),
+        _accessor_key(accessor_key) {}
+
+  // Return by reference as GuardAccessor owns the GuardManager.
+  std::unique_ptr<GuardManager>& get_guard_manager() {
+    return _guard_manager;
+  }
+
+  bool matches_key(const py::object key) const {
+    return _accessor_key.equal(key);
+  }
+
+  virtual ~GuardAccessor() = default;
+  virtual PyObject* access(PyObject* obj) const = 0;
+
+ private:
+  // Guard manager corresponding to the retrieved value from the GuardAccessor.
+  std::unique_ptr<GuardManager> _guard_manager;
+
+ protected:
+  // accessor key could be py::str for getattr, getitem or py::function for
+  // lambda accessor.
+  py::object _accessor_key;
+};
+
+/**
+ * Represents __getattr__ acccessor.
+ */
+class GetAttrGuardAccessor : public GuardAccessor {
+ public:
+  GetAttrGuardAccessor(py::str name)
+      : GuardAccessor(name), _attr_name(name.ptr()) {}
+
+  PyObject* access(PyObject* obj) const override { // borrowed ref
+    return PyObject_GetAttr(obj, _attr_name);
```

Comment:
`PyObject_GetAttr` returns a new reference. So, should we probably DECREF it somwhere?